### PR TITLE
chore: Move test requirements to optional dependencies in pyproject.toml

### PR DESCRIPTION
**Pull request by sbrugman**
## Description
Fixes open todos:
- Migrate testing requirements to optional dependencies
- Use fully static metadata for `kedro-datasets` (using [recursive optional dependencies](https://hynek.me/articles/python-recursive-optional-dependencies/))

## Development notes
Closes https://github.com/kedro-org/kedro/issues/1498 and closes https://github.com/kedro-org/kedro-plugins/issues/169

## Checklist

- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [X] Updated the documentation to reflect the code changes
- [X] Added a description of this change in the relevant `RELEASE.md` file
- [X] Added tests to cover my changes

### CODE REVIEW

**Review by astrojuanlu — COMMENTED** [review]
This is fantastic, thanks a lot @sbrugman!

The biggest change is going from our big dynamic `setup.py` in `kedro-datasets` to having all optional dependencies using self-referential requirements. I am personally a fan of this approach, but I am deferring to the engineering team to do the final approval.

**Comment by astrojuanlu** [comment]
e2e tests will need amendments:

```
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'test_requirements.txt'
```

**Review by merelcht — APPROVED** [review]
This is super neat! ⭐ Thank you so much @sbrugman ! I left one question, but other than that I'm super happy to get this merged.

File: kedro-airflow/features/environment.py [hunk 1]
```diff
@@ -50,8 +50,7 @@ def call(cmd, print_output=False):
 
     # pip install us
     call([context.python, "-m", "pip", "install", "-U", "pip", "pip-tools"])
-    call([context.pip, "install", "-r", "test_requirements.txt"])
-    call([context.pip, "install", "."])
+    call([context.pip, "install", ".[test]"])
```
[the remark this answers is not in the document]
Comment (sbrugman) [reply kedro-airflow/features/environment.py:53]: `call([context.pip, "install", ".[test]"])` is the same as `call([context.pip, "install", "."])` + `pip install -r test_requirements.txt`

[the remark this answers is not in the document]
Comment (sbrugman) [reply kedro-airflow/features/environment.py:53]: This works now that `test` is available in `optional-dependencies` in the `pyproject.toml` file

**Comment by sbrugman** [comment]
Rebased the changes on top of the `pre-commit`. 

@merelcht Could we merge this and then take on purging related dependencies from the core repo for the 0.19.0 release separately?

File: kedro-airflow/features/environment.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (merelcht) [reply kedro-airflow/features/environment.py:53]: Cool, thanks for explaining!

**Review by astrojuanlu — APPROVED** [review]

**Pull request merged**